### PR TITLE
Remove types deriving from JsonTypeInfo<T> and mark as sealed.

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Removes a number of generic types deriving from `JsonTypeInfo<T>` and marks the type itself as sealed. It also removes a number of abstract and virtual methods that were being used to perform delayed metadata resolution. The type itself should now only contain data and caches, with any metadata resolution logic being moved out to static factory methods.

Contributes to #80917.

cc @eerhardt

### CODE REVIEW

File: src/libraries/System.Text.Json/ref/System.Text.Json.cs [hunk 1]
```diff
@@ -1271,7 +1271,7 @@ public static partial class JsonTypeInfoResolver
     {
         public static System.Text.Json.Serialization.Metadata.IJsonTypeInfoResolver Combine(params System.Text.Json.Serialization.Metadata.IJsonTypeInfoResolver?[] resolvers) { throw null; }
     }
-    public abstract partial class JsonTypeInfo<T> : System.Text.Json.Serialization.Metadata.JsonTypeInfo
+    public sealed partial class JsonTypeInfo<T> : System.Text.Json.Serialization.Metadata.JsonTypeInfo
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/ref/System.Text.Json.cs:1274]: @stephentoub I'm making this change on the assumption that it isn't a breaking change. Is that correct?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs [hunk 2]
```diff
@@ -10,88 +10,87 @@
 namespace System.Text.Json.Serialization.Metadata
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/ReflectionJsonTypeInfoOfT.cs:10]: Note that I didn't rename either `ReflectionJsonTypeInfoOfT.cs` nor `SourceGenJsonTypeInfoOfT.cs` for fear of destroying the diffs. I'll follow up with a commit that renames these appropriately.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 3]
```diff
@@ -37,6 +35,38 @@ public abstract partial class JsonTypeInfo
         private Action<object>? _onDeserializing;
         private Action<object>? _onDeserialized;
 
+        internal JsonTypeInfo(Type type, JsonConverter converter, JsonSerializerOptions options)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:38]: Moved constructor to the top of the file.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 4]
```diff
@@ -1104,6 +1064,60 @@ internal static bool IsInvalidForSerialization(Type type)
             return type == typeof(void) || type.IsPointer || type.IsByRef || IsByRefLike(type) || type.ContainsGenericParameters;
         }
 
+        internal void PopulatePolymorphismMetadata()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:1067]: Methods moved from `JsonTypeInfo<T>`.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs [hunk 5]
```diff
@@ -0,0 +1,171 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs:4]: File moved from `SourceGenJsonTypeInfoOfT.cs`, see first commit for an easier diff.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 6]
```diff
@@ -37,6 +35,38 @@ public abstract partial class JsonTypeInfo
...
+            Options = options;
+            Converter = converter;
+            PropertyInfoForTypeInfo = CreatePropertyInfoForTypeInfo();
+            ElementType = converter.ElementType;
+
+            switch (converter.ConverterStrategy)
+            {
+                case ConverterStrategy.Dictionary:
+                    {
+                        KeyType = converter.KeyType;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:50]: could we just set it always similarly to ElementType and then get rid of switch statement which seems can be replaced with `if (converter.ConverterStrategy == ConverterStrategy.None)`?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs [hunk 7]
```diff
@@ -799,13 +805,14 @@ internal JsonTypeInfo JsonTypeInfo
...
-                Debug.Assert(_jsonTypeInfo == null || _jsonTypeInfo == value);
                 _jsonTypeInfo = value;
             }
         }
 
-        internal bool IsIgnored => _ignoreCondition == JsonIgnoreCondition.Always;
+        /// <summary>
+        /// Property was marked JsonIgnoreCondition.Always and also hasn't been configured by the user.
+        /// </summary>
+        internal bool IsIgnored => _ignoreCondition is JsonIgnoreCondition.Always && Get is null && Set is null;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs:815]: is this true even when there is custom converter?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs (hunk 6 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:50]: Looks like it. This switch statement used to be doing much more work in the past but could be removed now and replaced with the `GetTypeInfoKind` method.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+using System.Diagnostics;
+using System.Reflection;
+using System.Text.Json.Serialization.Converters;
+
+namespace System.Text.Json.Serialization.Metadata
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs:9]: nit: `namespace System.Text.Json.Serialization.Metadata;`

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonMetadataServices.Helpers.cs:9]: Would be nice to save ourselves a bit of indentation, but I don't think we use the new namespace syntax anywhere in libraries at the moment. If we do, we should make sure it is applied consistently in a dedicated PR.

@stephentoub do we have any plans/concerns about doing that in libraries eventually?

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs [hunk 9]
```diff
@@ -206,15 +216,21 @@ public IList<JsonPropertyInfo> Properties
         {
             get
             {
-                if (_properties == null)
-                {
-                    PopulatePropertyList();
-                }
-
-                return _properties;
+                _isUserModifiedPropertyList = true;
```
Comment (krwq) [inline src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:219]: won't this always be true now?

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs:219]: Yes, assuming the user calls the property getter.

File: src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs (hunk 7 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs:815]: If you're asking about the converter on the property type, then yes it should hold irrespective of that.

**Review by krwq — APPROVED** [review]

**Pull request merged**